sources/niUI/src/VideoDecoderTheoraPlayer.cpp

```diff
@@ -25,70 +25,60 @@ class cVideoDecoderTheoraPlayer : public ni::cIUnknownImpl<ni::iVideoDecoder,ni:
   {
     mfSpeed = 1.0f;
     mbPaused = eFalse;
-    mbDecoding = eFalse;
     mnNumLoops = eInvalidHandle;
     mnLastUpdateFrameId = eInvalidHandle;
     mptrGraphics = apGraphics;
     mpTargetTexture = niNew sVideoDecoderTexture(apGraphics,ahspName,this,NULL);
     niLog(Info,niFmt("Video started loading: '%s'",apFile->GetSourcePath()));
 
     const tBool withAlpha = StrEndsWith(apFile->GetSourcePath(),".ogw");
-
-    iExecutor* pLoadingExecutor =
-        ni::GetConcurrent()->GetExecutorIO()
-        // ni::GetConcurrent()->GetExecutorMain()
-        ;
-
     Ptr<iFile> _fp = apFile;
-    Ptr<cVideoDecoderTheoraPlayer> _this = this;
-    niExec_(pLoadingExecutor,_this,_fp,withAlpha) {
-      TheoraVideoClip* pTheoraVideoClip = Theora_createVideoClip(
-          _fp,
-          withAlpha ? VIDEO_OUTPUTMODE_ALPHA : VIDEO_OUTPUTMODE,
-          2,
-          false);
-      if (!pTheoraVideoClip) {
-        niError(niFmt("Can't create the video clip '%s'.",_fp->GetSourcePath()));
-        return ni::eFalse;
-      }
+    mpTheoraVideoClip = Theora_createVideoClip(
+      _fp,
+      withAlpha ? VIDEO_OUTPUTMODE_ALPHA : VIDEO_OUTPUTMODE,
+      4,
+      false,
+      false);
+
+    if (!mpTheoraVideoClip) {
+      niError(niFmt("Can't create the video clip '%s'.",_fp->GetSourcePath()));
+      return;
+    }
 
-      niExec(Main,_this,pTheoraVideoClip,_fp,withAlpha)  {
-        _this->mpTheoraVideoClip = pTheoraVideoClip;
+    if (!mpTheoraVideoClip->load(_fp)) {
+      niError("Couldn't load the video clip.");
+      return;
+    }
 
-        if (niFlagIs(_this->mFlags,eVideoDecoderFlags_TargetTexture)) {
-          Ptr<iTexture> ptrTex = _this->mptrGraphics->CreateTexture(
-              NULL,
-              eBitmapType_2D,
-              VIDEO_PIXELFORMAT,
-              0,
-              pTheoraVideoClip->getWidth(), pTheoraVideoClip->getHeight(), 0,
-              eTextureFlags_Dynamic);
-          if (!ptrTex.IsOK()) {
-            niError(niFmt(_A("Can't create target texture for video '%s'."),_fp->GetSourcePath()));
-            return ni::eFalse;
-          }
-          _this->mpTargetTexture->mptrTexture = ptrTex;
-        }
+    if (niFlagIs(mFlags,eVideoDecoderFlags_TargetTexture)) {
+      Ptr<iTexture> ptrTex = mptrGraphics->CreateTexture(
+        NULL,
+        eBitmapType_2D,
+        VIDEO_PIXELFORMAT,
+        0,
+        mpTheoraVideoClip->getWidth(), mpTheoraVideoClip->getHeight(), 0,
+        ni::eTextureFlags_Dynamic|ni::eTextureFlags_Overlay);
```

Comment:
Should probably be an overlay as an option. But keep in mind this is meant for ingame animated textures so they are supposed to be pow2, etc... The issue with Overlay textures is that they are not allowed to repeat and have some other limitiations on some hardware, thats why its a flag versus the default.

Ideally we'd add video creation flags to handle this - aswell as "decodeFirstFrame".

Reply:
without ni::eTextureFlags_Overlay flag, the texture will resized to pow of 2, it will cause a resize when upload texture...

Reply:
just this one alone, will speed up the video decoding a lot